### PR TITLE
Add parallelization framework

### PR DESCRIPTION
**Pull request by sdrave**
The pull request contains pyMOR's parallelization framework from the pymor-paper branch (only the framework, not the parallelized algorithms).

Before we merge to master, we should double-check if we are satisfied with the interfaces. (A bit of documentation would also be nice, I guess.)

As a first commit I would propose to rename some methods as follows:
- distribute -> push
- distribute_array -> scatter_array
- distribute_list -> scatter_list

Any comments?

### CODE REVIEW

**Comment by renemilk** [comment]
Pro renaming to the standard "scatter", but why the mismatch with push?

**Comment by sdrave** [comment]
Different semantics: push does not scatter but copy the same data to each worker.

**Comment by renemilk** [comment]
Ah, +2 then.

File: src/pymor/parallel/ipython.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+try:
+    from ipyparallel import Client, TimeoutError
+    HAVE_IPYTHON = True
+except ImportError:
+    try:
+        from IPython.parallel import Client, TimeoutError
+        HAVE_IPYTHON = True
+    except ImportError:
+        HAVE_IPYTHON = False
```
Comment (ftalbrecht) [inline src/pymor/parallel/ipython.py]: It seems like `HAVE_IPYTHON` is not used within this file. Do we require it somewhere else?

Comment (sdrave) [inline src/pymor/parallel/ipython.py]: I do not think so. I would like to keep it, however, as we have the same flags for other optional dependencies.

**Comment by sdrave** [comment]
I have just updated the names of the interface methods.

**Comment by sdrave** [comment]
Another thing:
Right now, `push`, `scatter_array`, `scatter_lists` return context managers. Their `__enter__` method then returns the remote object handles. For instance, I can write
```python
with pool.push(d) as rd:
    pool.apply(function, d=rd)
```

Doing so ensures that remote resources are always freed when we leave the context. On the downside, these methods are a real PITA to use in interactive session or more complex scenarios, when you cannot easily put everything inside one block.

I would propose to make these methods behave more like Python's `open`: They directly return the remote object handle (directly transmitting the data) and this handle can also be used as a context manager. `__enter__` will do nothing whereas `__exit__` will call the object's `destroy` (`free`, `remove`?) method. 

There is only one downside: Right now, you can write:
```python
with pool.push(d, U) as rd, rU:
    do something
```
which will probably become impossible. Instead you would have to write
```python
with pool.push(d) as rd, pool.push(U) as rU:
    do something
```
which is not only more verbose but also potentially more expensive. However, I believe this is acceptable. (In case you have really lots of objects to push, you can put them all into a single `dict` which is then pushed.)

**Comment by renemilk** [comment]
> However, I believe this is acceptable

Agreed.

**Comment by ftalbrecht** [comment]
Btw, the reported [reimport problem](https://landscape.io/diff/217578) looks legit to me...

**Comment by sdrave** [comment]
@ftalbrecht, as far as I remember, this is required. (This is executed on the workers and strange things happen there.)

**Comment by sdrave** [comment]
Ok, so I changed the behaviour of the push and scatter methods and also added a `DummyLogger`.

**Comment by sdrave** [comment]
Finally, I have added a simple `RemoteObjectManager`. See #155 for usage. From my point of view, what is left to do is to add some documentation.

**Comment by ftalbrecht** [comment]
Great! I will not be able to do any substantial work (on this) until end of october...

**Comment by sdrave** [comment]
Finally, I have added some documentation. Please have a look, if the interface seems halfway clearly defined. If yes, I will merge to master.

**Comment by ftalbrecht** [comment]
Thank you for the documentation, very nice! What came to my mind: could it be desirable to be able to provide a weighting of the arguments for the `scatter_...` functions?

**Comment by sdrave** [comment]
What do you mean by 'weighting'?

**Comment by ftalbrecht** [comment]
Well, the even scattering of data to the workers assumes that all elements induce the same amount of work and that all workers are evenly potent. I was just wondering whether one would like to scatter elements not-evenly, given some a priori knowledge (e.g., the first element of the list will induce twice as much work as the other elements); this would correspond to a weighting of the elements to be scattered, which would then be taken into account when splitting the arrays or lists...

**Comment by sdrave** [comment]
Ah, I see. This feels a bit to complicated to me. Moreover, you can always easily do the partitioning yourself.

**Comment by ftalbrecht** [comment]
Yes, I agree; it was just an idea I wanted to bring up, before we
finalize the interfaces.

No objections against merging this with master...

**Pull request merged**